### ISSUE TITLE
Simplify consumption of C# centric object builders

### ISSUE THREAD

**Issue by wallymathieu** [fsharp/fslang-suggestions#877]
# Simplify consumption of C# centric object builders

I propose we add support in some form to allow the construction of objects in line with how API:s are defined for use in C#

One alternative that might go with the grain of current usage of computation expression builders would be the possibility of a computation expression wrapper that allows for constructing mutable objects in a similar fashion to how they are constructed in C# syntactic sugar for [object and collection initializers](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers). Similar functionality exists within [VB.net](https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/objects-and-classes/object-initializers-named-and-anonymous-types), but in a manner more appropriate for that style of programming.

The existing way of approaching this problem in F# is to do something similar as:

```f#
let v = ConfigurationBuilder(UseExplicitWhatnot = true)
v.Sublevel.BasementMapping <- mapping
v.ListOfFloors.Add floor1
...
```

that's fine for small usages of C# API:s, but can be a bit annoying when trying to consume larger chunks.

## Pros and Cons 

The advantages of making this adjustment to F# are to make it more compatible with C# intended API:s

The disadvantages of making this adjustment to F# are that it could complicate the F# compiler.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): 

I'm unsure about the cost since this is sort of a rough suggestion

Related suggestions: (put links to related suggestions here)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [ ] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by wallymathieu** [comment]
After asking around a bit, this might be due to my confusion around object initialization in f#

**Comment by wallymathieu** [comment]
I've seen a lot of code in C# where you do:
```c#
var o = new Ctor{
 OtherProperty={
   Property1={
     {Key, Value}
   },
   Property2={
     Property3=Value3
   }
 }
} 
```
perhaps there is some way to do the above in f# as well

**Comment by wallymathieu** [comment]
```c#
using System;
using Xunit;
namespace cstesting
{
    public class Type1{
        public Type2 Property{get;}=new Type2();
    }
    public class Type2{
        public int MyProperty { get; set; }
    }
    public class UnitTest1
    {
        [Fact]
        public void Test1()
        {
            var ugly= new Type1{ Property={ MyProperty=1 } };
        }
    }
}
```

**Comment by wallymathieu** [comment]
Object construction for known types with c# centric behavior

**Comment by abelbraaksma** [comment]
> c# centric behavior

I just don't understand this phrase, behavior is equal in all .NET languages. But since you closed this, I assume you've found a solution.

**Comment by TheJayMann** [comment]
From what I understand, what is being asked is feature parity with C# in terms of object initializers and collection initializers.  Currently, F# does support object initializers, but it only allows assignment directly to the property, and F# does not support collection initializers outside of array creation, F# list creation, and sequence computation expression generators.  C# object initializers will also allow directly setting the property of an object which itself is the property of the object being instantiated if the constructor initializes the property to a not null value, whereas it appears F# object initializers can only set the top level properties to a new value.  Also, C# collection initializers involve calling a defined `Add` method (and I believe `Add` extension methods can be used) to add values to a collection being created, or to a collection already instantiated on a property of an object being initialized.  I have an example of both cases of C# object initialization and collection initialization which I do not believe is possible with F# initializers.
```csharp
class NestedClass { 
    public int Num { get; set; }
}
class RootClass {
    public NestedClass Nested { get; } = new NestedClass();
    public List<int> NumList { get; } = new List<int>();
}
new RootClass() { 
    Nested = {
        Num = 42
    },
    NumList = {4,8,15,16,23,42},
}
```
The result for C# interactive:
```
Submission#1.RootClass { Nested=Submission#0.NestedClass { Num=42 }, NumList=List<int>(6) { 4, 8, 15, 16, 23, 42 } }
``` 

From what I have tested in F# interactive, initializing the `Nested` property using an F# object initializer would only work if the `Nested` property were read/write, and only if it the initializer itself were creating a new `NestedClass` directly.  Also, for the `NumList` property, again, would only work if the property were read/write, as well as having to create an array, F# list, or sequence expression, and then creating a new `ResizeArray` passing the previously generated sequence, while not being able to make use of the collection object which was already created by the object during construction.

These are the arguments being made in so far as I have come to understand them.

**Comment by abelbraaksma** [comment]
@TheJayMann, thanks it makes sense now. I wasn't aware of these differences, guess I haven't come across them in real life yet.

Sounds like a reasonable request, not sure why it was closed.

**Comment by baronfel** [comment]
I think the biggest gap is the missing collction-initializer syntax, right? That would need a type-directed syntax rule to allow for providing values of type `'t` where there exists a member `Add('t)` on the source collection type. Then you start to think about if mutable addition vs immutable addition should be supported, meaning if the return value of Add is unit or 'collection.

I think it makes sense, but I wonder how it would play with the existing list/array initialization syntax.

**Comment by wallymathieu** [comment]
It's nested initialization and collection initialization syntax sugar. I closed it because it might be good enough. If we reformulate the issue I have no problem with reopening it.

**Comment by wallymathieu** [comment]
Should I reopen it @baronfel ?

**Comment by baronfel** [comment]
I think it's worth it to, the gap with the Add syntax is a pretty common request for beginners from c# in my opinion.

**Comment by wallymathieu** [comment]
Yes, I agree on the add syntax. How would it make sense from a F# perspective? Something like:
```f#
let v = ConfigurationBuilder(UseExplicitWhatnot = true, ListOfFloors @= [floor1])
```
?

**Comment by dsyme** [comment]
Tracked by #969
